src/java/org/apache/cassandra/service/snapshot/SnapshotManager.java

```diff
@@ -17,156 +17,830 @@
+     * @param olderThanTimestamp clear the snapshot if it is older than given timestamp
+     * @param includeEphemeral   whether to include ephemeral snapshots as well
+     * @return predicate which filters snapshots on given parameters
+     */
+    static Predicate<TableSnapshot> shouldClearSnapshot(String tag,
+                                                        Set<String> keyspaces,
+                                                        long olderThanTimestamp,
+                                                        boolean includeEphemeral)
+    {
+        return ts ->
         {
-            if (!expiredSnapshot.isExpired(now()))
-                break; // the earliest expiring snapshot is not expired yet, so there is no more expired snapshots to remove
+            // When no tag is supplied, all snapshots must be cleared
+            boolean clearAll = tag == null || tag.isEmpty();
+            if (!clearAll && ts.isEphemeral() && !includeEphemeral)
+                logger.info("Skipping deletion of ephemeral snapshot '{}' in keyspace {}. " +
+                            "Ephemeral snapshots are not removable by a user.",
+                            tag, ts.getKeyspaceName());
+            boolean passedEphemeralTest = !ts.isEphemeral() || (ts.isEphemeral() && includeEphemeral);
+            boolean shouldClearTag = clearAll || ts.getTag().equals(tag);
+            boolean byTimestamp = true;
 
-            logger.debug("Removing expired snapshot {}.", expiredSnapshot);
-            clearSnapshot(expiredSnapshot);
-        }
+            if (olderThanTimestamp > 0L)
+            {
+                Instant createdAt = ts.getCreatedAt();
+                if (createdAt != null)
+                    byTimestamp = createdAt.isBefore(Instant.ofEpochMilli(olderThanTimestamp));
+            }
+
+            boolean byKeyspace = (keyspaces.isEmpty() || keyspaces.contains(ts.getKeyspaceName()));
+
+            return passedEphemeralTest && shouldClearTag && byTimestamp && byKeyspace;
+        };
     }
 
     /**
-     * Deletes snapshot and remove it from manager
+     * Takes a snapshot by creating hardlinks into snapshot directories. This method also
+     * creates manifests and schema files and such snapshot will be added among tracked ones in this manager.
+     *
+     * @param cfs          column family to create a snapshot for
+     * @param tag          name of snapshot
+     * @param ephemeral    true if the snapshot is ephemeral, false otherwise
+     * @param ttl          time after the created snapshot will be removed
+     * @param creationTime time the snapshot was created
+     * @param rateLimiter  limiter for hard-links creation, if null, limiter from DatabaseDescriptor will be used
+     * @return logical representation of a snapshot
      */
-    public synchronized void clearSnapshot(TableSnapshot snapshot)
+    public TableSnapshot createSnapshot(ColumnFamilyStore cfs,
```

Comment:
One of the goals of centralizing snapshot logic into `SnapshotManager` was to simplify things but this snapshot creation logic is getting complex enough that probably deserves its own class (ie. `CreateSnapshotTask`) that receives a set of snapshot creation options (ie. tag/ttl/etc) and returns a `Set<TableSnapshot>` that is then added to the set of `liveSnapshots`. 

This would remove the coupling of `SnapshotManager` with `ColumnFamilyStore` since all the handling of snapshot creation would be encapsulated in the `CreateSnapshotTask`.

This would make `SnapshotManager` be responsible for the management of snapshot lifecycle (similar to `CompactionManager`) while the actual snapshot creation task would be encapsulated in its own `CreateSnapshotTask` class (similar to `AbstractCompactionTask`). This is analogous to encapsulating snapshot listing logic into its own `SnapshotLoader` class to make `SnapshotManager` leaner.

With this the implementation of `SnapshotManagerMBean.takeSnapshot` would be something like:
```java
    synchronized takeSnapshot(String tag, Map<String, String> options, String... entities)
   {
          CreateSnapshotTask task = new CreateSnapshotTask(tag, entitites, snapshotRateLimiter);
          Set<TableSnapshot> snapshot = task.executeBlocking()
          liveSnapshots.addAll(snapshot)
   } 
```

Do you think this makes sense ?

Reply:
it is better to base it on a builder pattern because I hit a situation when I had "telescopic constructors" and creating a builder for that was just more comfortable.